src/main/java/com/iota/iri/controllers/MilestoneViewModel.java

```diff
@@ -95,23 +108,39 @@ public MilestoneViewModel next(Tangle tangle) throws Exception {
     }
 
     public static MilestoneViewModel findClosestPrevMilestone(Tangle tangle, int index) throws Exception {
-        Pair<Indexable, Persistable> milestonePair = tangle.previous(Milestone.class, new IntegerIndex(index));
-        if(milestonePair != null && milestonePair.hi != null) {
-            return new MilestoneViewModel((Milestone) milestonePair.hi);
+        // search for the previous milestone preceding our index
+        MilestoneViewModel previousMilestoneViewModel = null;
+        int currentIndex = index;
+        while(previousMilestoneViewModel == null && --currentIndex >= index - MAX_MILESTONE_INDEX_GAP) {
+            previousMilestoneViewModel = MilestoneViewModel.get(tangle, currentIndex);
         }
-        return null;
+
+        return previousMilestoneViewModel;
     }
 
-    public static MilestoneViewModel findClosestNextMilestone(Tangle tangle, int index, boolean testnet,
-                                                              int milestoneStartIndex) throws Exception {
-        if(!testnet && index <= milestoneStartIndex) {
-            return first(tangle);
```

Comment:
I don't get the question.

This just removes the ability to provide a completely erroneous milestone index in the call of the method. In fact all parts of the code that call this method use milestone indexes that are by definition bigger or equals to the milestoneStartIndex that is provided via the config. So its a case that never ever happens (not even in tests).

I think it is better if the method would tell me that it couldnt find a milestone for the given index rather than magically "repair" my parameters. In addition the "first" element in the database is not the first milestone. If you for example start syncing a node from the scratch a week after the global snapshot the first milestone you will receive is not the first milestone issued by the coordinator but one that is 1 week to old so this "param repair" does more harm than good.

Reply:
Hmm,
Have a look at [EntryPointSelectorImpl](https://github.com/iotaledger/iri/blob/dev/src/main/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorImpl.java#L34-L38). It seems that if the user via the API will use a large enough depth then you will default on the `latestSolidSubtangleMilestone`, which is the opposite of what the user expected.

The previous behavior in this case makes much more sense, since the user wanted a very early milestone.

Anyhows, if you are still  unhappy with this behavior feel free to open an issue about it, and it can be discussed and handled in a separate PR. Please revert to the previous behavior.

Reply:
The latest changes of the findClosestNextMilestone method will now also handle indexes that are too small by just iterating towards it (i removed the "automatic cutoff point" of the MAX_..._GAP and introduced a cutoff point in the parameter which can then be controlled by the caller) since it was network specific anyway.

If i really want to find the exact successor i can just use index + 1 when using MilestoneViewModel.get() and don't use this method.

So now this method will do exactly what it says - find the next milestoneViewModel in the database that follows after the given index (even if the index starts out to be smaller than the milestoneStartIndex.

Reply:
Ok,
@alon-e and I investigated what is going on in this code yesterday. Apparently the removal of the `testnet` flag solves #637. Can you please edit the description of the PR to mention this?